### ISSUE TITLE
`jj untrack foo~` fails

### ISSUE THREAD

**Issue by 0-wiz-0** [martinvonz/jj#4509]
## Description

emacs writes backup files with a `~` (tilde) at the end. jj automatically picks them up if you forgot to ignore them in the `.gitignore` file.
I can't untrack them in the latest version any longer. (I think this worked when it was `jj untrack`, but now it doesn't with `jj file untrack`, but perhaps I misremember.)

## Steps to Reproduce the Problem

1. jj git init
1. touch foo~
1. jj
1. jj file untrack foo~
```
jj file untrack foo\~
Error: Failed to parse fileset: Syntax error
Caused by:  --> 1:5
  |
1 | foo~
  |     ^---
  |
  = expected `~` or <primary>
```

## Specifications

- Platform: NetBSD/10.99.12/amd64
- Version: 0.21.0

**Comment by yuja** [comment]
Since fileset is enabled by default, `~` has to be quoted `'"foo~"'` or `'glob:"*~"'`.
https://martinvonz.github.io/jj/latest/filesets/

Maybe we can add a hint if the expression looks like an existing path.

**Comment by 0-wiz-0** [comment]
From a quick read of the linked page: `~` seems only to be supported as a separate word, or at the start of a word, so perhaps making that an explicit requirement would also solve this.

**Comment by yuja** [comment]
Yeah, trailing `~` could be special cased, but I don't think it's worth the added oddity. `*~` is usually a backup file and so often ignored by `.gitignore` or user ignore.

`x~y` and `~x` is a valid syntax, btw.

**Comment by samueltardieu** [comment]
The message is now:

```
Error: Failed to parse fileset: Syntax error
Caused by:  --> 1:5
  |
1 | foo~
  |     ^---
  |
  = expected `~` or <primary>
Hint: See https://martinvonz.github.io/jj/latest/filesets/ for filesets syntax, or for how to match file paths.
```

Hopefully this is a clearer message. If this is, might we close this issue?

**Comment by 0-wiz-0** [comment]
I accept that I'm the only one annoyed by the change from 'jj untrack' to 'jj file untrack' and have worked around it by having my global git config ignore "*~" files.

As for the actual change - I think this needs improvements in the documentation.
I read `expected `~` or <primary>` as that `jj` would expect a second tilde or a <primary> after the first tilde.
Both alternatives are unclear to me.
Neither https://martinvonz.github.io/jj/latest/filesets/ nor the page it links to from 'must be quoted', https://martinvonz.github.io/jj/latest/templates/#string-literals , mention `~~`, and in my reading of https://martinvonz.github.io/jj/latest/templates/#string-literals I don't see a mention of how I should quote `~` at all, since the full list of backslash-quoted characters doesn't include the tilde 
The word primary is not mentioned in the filesets page at all, so I don't know what a `<primary>` is either.

**Comment by martinvonz** [comment]
> the change in behaviour between the old 'jj untrack' and the new 'jj file untrack'

FYI, I don't think it's a difference between `jj untrack` and `jj file untrack`. `jj untrack` actually still exists as an alias. What changed is that we enabled filesets by default.

> I read `expected `~` or <primary>` as that `jj` would expect a second tilde or a `<primary>` after the first tilde.

`~` is simply an operator for subtraction/difference (that's why it looks similar to `-`, which we chose not to use because it's so common in filenames).

A second tilde followed by an expression would make the syntax correct, but it's probably not what. The error message is generated by the low-level parsing code without understanding what makes a useful expression. `foo~~bar` means "file `foo` but not (not file `bar`)`, i.e. an empty set.

I think you can think of a `<primary>` is a fileset expression.

> Neither https://martinvonz.github.io/jj/latest/filesets/ nor the page it links to from 'must be quoted',

That page includes this:

> File names passed to these commands [must be quoted](https://martinvonz.github.io/jj/latest/templates/#string-literals) if they contain whitespace or meta characters.

In `foo~`, the `~` is such a "meta character". @yuja, should we simply call it "operator" for simplicity? I'm guessing "meta characters" also includes the `:` in `glob:` etc., but do we need make that distinction? If we do, we should probably document it.

> As for the actual change - I think this needs improvements in the documentation.

I'm not arguing with this. The above is just to explain how it actually works.

**Comment by ilyagr** [comment]
> Maybe we can add a hint if the expression looks like an existing path.

This sounds like a good idea to me, it's consistent with what we try to do in other similar cases. I just wanted to resurface it so that it doesn't get lost in the discussion.

**Comment by yuja** [comment]
> In `foo~`, the `~` is such a "meta character". @yuja, should we simply call it "operator" for simplicity? I'm guessing "meta characters" also includes the `:` in `glob:` etc., but do we need make that distinction?

I just thought "meta character" is more abstract and easier to imagine than "operator". If we need a strict definition, maybe we can add a link to the grammar file?

> > Maybe we can add a hint if the expression looks like an existing path.
> 
> This sounds like a good idea to me, it's consistent with what we try to do in other similar cases.

I thought this would be a nice addition, but I think the current state (a hint with doc URL) is pretty good. We can improve the doc by adding more bad/good examples. (If I were new to jj, I would probably search the doc by error message.)

**Comment by taranlu-houzz** [comment]
Hello, I am running into a similar issue with a file that got accidentally tracked that has a single quote: `My Test's test.xlsx` and I cannot seem to get `jj` to untrack the file. I've tried different variations of quoting and escaping characters, but nothing has worked for me and I always get this error:
```
❯ jj file untrack "My Test's test.xlsx"
Error: Failed to parse fileset: Syntax error
Caused by:  --> 1:8
  |
1 | My Test's test.xlsx
  |        ^---
  |
  = expected <EOI>
```

**Comment by martinvonz** [comment]
The quoting depends on your shell. Which shell do you use? I think in Bash, this would work:`jj file untrack '"My Test'\''s test.xlsx"'`

**Comment by taranlu-houzz** [comment]
@martinvonz I was just about to update my comment with that. I use `fish`, and that is exactly what ended up working.

**Comment by scott2000** [comment]
Maybe it would be a good idea to add a global `--raw-paths` option which disables fileset parsing for command line arguments. Then the error message could give a simple solution without having to explain quoting, like "This argument is being interpreted as a fileset. If you want to use a plain file path, consider using \`--raw-paths\` to prevent it from being interpreted as a fileset."

This would probably also be good for shell scripts to avoid having to escape paths. It's a bit different from the current `ui.allow-filesets` option, since it would only apply to command line arguments, not filesets in other places.

**Comment by ilyagr** [comment]
For reference, there was a [related discussion on Discord](https://discord.com/channels/968932220549103686/968932220549103689/1304563334464409654) (+ [thread](https://discord.com/channels/968932220549103686/1304567372073402388) where somebody encountered this confusing error:

```console
$ jj restore emacs/.emacs.d/lisp/\#init-os.el\# 
Error: Failed to parse fileset: Syntax error
Caused by:  --> 1:21
  |
1 | emacs/.emacs.d/lisp/#init-os.el#
  |                     ^---
  |
  = expected <EOI>, `|`, `&`, or `~`
Hint: See https://martinvonz.github.io/jj/latest/filesets/ for filesets syntax, or for how to match file paths.
```

@bnjmnt4n found a related TODO:

https://github.com/martinvonz/jj/blob/528ccb318e084d10d4951ef976b3c0b1729e0413/lib/src/fileset.pest#L30-L36

**Comment by yuja** [comment]
> Maybe it would be a good idea to add a global `--raw-paths` option which disables fileset parsing for command line arguments.

If we need an escape hatch, a modifier syntax might be better. Something like `raw:<PATH>` or `literal:<PATH>`.

For this emacs example, we can add `#` to the base string characters.

**Comment by ilyagr** [comment]
There was some discussion in #5517, but I think everybody seems happy to try Yuya's idea from the previous comment.

Perhaps the hint from #5517 could then say something "For a literal filename, try prefixing it with `literal:`".

--------------------

If that is somehow not enough, I wanted to record in this thread that we could consider making only `:` a special symbol by default and assume that something is a literal path unless it starts with `exact:`, `prefix:`, or something else to specify that it's a fileset (we could also have `set:` if needed).

I'm personally also still curious about https://github.com/jj-vcs/jj/issues/4509#issuecomment-2398938140 and https://github.com/jj-vcs/jj/issues/4509#issuecomment-2398952703 . Neither of these conflicts with Yuya's idea, but they might also not be necessary.

**Comment by dzaima** [comment]
A downside of `literal:`/`raw:`/etc is that it's largely unusable by shell-expanded things, i.e. `jj file untrack raw:*~` or `function untrack-raw() { jj file untrack "raw:$@"; }; untrack-raw *~` wouldn't work.

**Comment by ViggoV** [comment]
I am seeing what I assume is the same issue with parentheses. NextJS makes heavy use of special character in folder names for routing, such as `/[...]/` for params, `/(...)/` for a group of routes and `/@.../` for parallel routes.

**Comment by spectral54** [comment]
We just had a user encounter this with `jj file list`, and we didn't get a nice error message when we encountered it, just:

```
$ jj file list foo~bar
Warning: No matching entries for paths: foo, bar
```

Is there a way of disabling all fileset operators unless there's some prefix? I'm thinking something like what Mercurial has where a `set:` prefix (I think that was it) was necessary to get filesets behavior. I get that it's more typing for people who want to use filesets, but if they're familiar with filesets, they can set the option to make them on by default. I think the default for users should be the least surprising and most compatible thing, which is raw files (not filesets, not globs (except maybe on Windows?))

**Comment by martinvonz** [comment]
Perhaps we could show a slightly different warning and a hint if the fileset is not just a union of explicit paths?

```
$  jj file list foo bar
Warning: No matching entries for paths: foo, bar
$ jj file list foo~bar
Warning: No matching entries for paths in fileset: foo, bar
Hint: See https://docs.jj-vcs.dev/latest/filesets/
```

The simplest way of implementing that would not print the hint for `foo|bar` because that would get parsed to a union of explicit paths. Seems good enough to me, especially because `|` is very uncommon in file names.

**Comment by yuja** [comment]
Maybe we can remove infix `~` if this is a common case of the problem? `x~y` is just a short for `x&~y`, and it's unlikely that a path includes `|` or `&`.

**Comment by CGamesPlay** [comment]
Just was bitten by this.

```
Modified files in working copy:
  M src/pages/gallery/[slug]/[...path].ts
  M src/pages/gallery/[slug].ts

$ jj restore "src/pages/gallery/[slug]/"
Nothing changed.
```

Is it reasonable that if the user gives an actual literal path that exists, that we don't try to interpret it as a something else?

**Comment by tim-janik** [comment]
> Maybe it would be a good idea to add a global `--raw-paths` option which disables fileset parsing for command line arguments.

There is a widely established unix convention for this already:
```sh
$ rm -f -- --wierd\*file\"with\'special\\chars
$ touch -- --wierd\*file\"with\'special\\chars
$ ls    -- --wierd\*file\"with\'special\\chars
'--wierd*file"with'\''special\chars'
$ jj file track -- --wierd\*file\"with\'special\\chars
Error: Failed to parse fileset: Syntax error
$ jj file track -- file:--wierd\*file\"with\'special\\chars
Error: Failed to parse fileset: Syntax error
```

IMHO, the above jj commands should just work correctly by treating everything after `--` as raw paths.
**Not** doing this makes scripting jj commands almost impossible (implementing jj expression quoting in the shell is hard and certainly not viable in a `for f in *; do ... done` loop) and it violates the principle of least surprise for anyone familiar with other unix commands.
FWIW, this also makes it hard to tell agents howto use jj commands, directly or in scripts.

**Comment by jennings** [comment]
Would it make sense to have a config option that makes filesets interpreted as literal paths by default instead of as glob patterns? Doing so could make this usable from any command accepting filesets without needing to add an option everywhere.

```command
# the -- prevents interpreting as an option, but the config is still
# needed to make literals the default pattern type
jj --config 'ui.default-fileset-pattern="file"' file track -- --weird\*file\"with\'special\\chars

# same as (if I got it right):
jj file track -- file:\"--weird\*file\\\"with\'special\\chars\"
```

(the `\'` is required in both cases for the shell)

**Comment by tim-janik** [comment]
> Would it make sense to have a config option that makes filesets interpreted as literal paths by default instead of as glob patterns? Doing so could make this usable from any command accepting filesets without needing to add an option everywhere.

> jj --config 'ui.default-fileset-pattern="file"' file track -- --weird\*file\"with\'special\\chars

This would still be very tedious in scripting and generic adhoc use like shell loops. 
While a config option would be nice, adding a simple CLI option that's supported by all commands taking a fileset would be nice, i.e.

` jj file track --raw -- --wierd\*file\"with\'special\\chars `

**Comment by josephlou5** [comment]
> Would it make sense to have a config option that makes filesets interpreted as literal paths by default instead of as glob patterns?

I've never used it (removed before I started using `jj`), but seems like `ui.allow-filesets` used to do this? It was removed in 07c63ed182bb1cbd9b52fe8e4f41638bdb5aafb6 though.

**Comment by yuja** [comment]
> IMHO, the above jj commands should just work correctly by treating everything after `--` as raw paths.

`--` is a separator to prevent following arguments from being parsed as named options. It shouldn't change how positional arguments are interpreted.

> > Would it make sense to have a config option that makes filesets interpreted as literal paths by default instead of as glob patterns?
> 
> I've never used it (removed before I started using `jj`), but seems like `ui.allow-filesets` used to do this?

It's a migration option. As far as I can tell, we avoid adding config knobs that significantly change the CLI behavior. The `raw:<PATH>` prefix is probably the easiest workaround. `--raw-path <PATH>` might also work.

**Comment by tim-janik** [comment]
> It's a migration option. As far as I can tell, we avoid adding config knobs that significantly change the CLI behavior. The `raw:<PATH>` prefix is probably the easiest workaround. `--raw-path <PATH>` might also work.

Please make sure this is a single switch not a per file option, so the following can work without per file gymnastics:

```sh
jj file track --raw-path -- *
```

**Comment by yuja** [comment]
> Please make sure this is a single switch not a per file option, so the following can work without per file gymnastics:
> 
> jj file track --raw-path -- *

Can you elaborate on why? For scripting use cases, `--raw-path=<PATH>` should just work. It's easy to translate each path argument. For interactive use, `jj file track '*'` can be used instead of letting the shell expand `*`.

Oh, and we'll need to decide if `--raw-path=<PATH>` or `raw:<PATH>` means `cwd:quote(<PATH>)`, `root:quote(<PATH>)`, or `quote(<PATH>)`. For scripting, `root:quote(<PATH>)` is probably the most convenient, but `quote(<PATH>)` works better in interactive use.

**Comment by tim-janik** [comment]
> > Please make sure this is a single switch not a per file option, so the following can work without per file gymnastics:
> > jj file track --raw-path -- *
> 
> Can you elaborate on why? For scripting use cases, `--raw-path=<PATH>` should just work. It's easy to translate each path argument. 

No, raw:* or --raw-path=* doesn't work with multiple arguments, and those can come from a shell variables ${FILES[@]} or from xargs or GNU Parallel, or a myriad of other commands.
Having to prefix each one is extremely tedious, it requires encoding the JJ fileset syntax escaping requirements into whatever tool you happen to use at the time.

> For interactive use, `jj file track '*'` can be used instead of letting the shell expand `*`.

No, `echo *` is just the most trivial example possible, I used that for brevity in Github comments. Other use casses: 

```sh
find -name \*.cc -o -name \*.hh -maxdepth 2 | xargs jj file track --raw --
```

The above becomes hard when any kind of per file handling is required, it becomes close to impossible when template syntax specific escaping rules must be followed.

> Oh, and we'll need to decide if `--raw-path=<PATH>` or `raw:<PATH>` means `cwd:quote(<PATH>)`, `root:quote(<PATH>)`, or `quote(<PATH>)`. For scripting, `root:quote(<PATH>)` is probably the most convenient, but `quote(<PATH>)` works better in interactive use.

No, just stick to what *every* other unix tool does.
The implementation should be this:

```sh
# jj file track --raw -- *
jj_file_track_raw() ( for f in "$@" ; do f=$(readlink -f -- "$f") && jj file track -- "file:$f" ; done )
```

Except for the `Error: Failed to parse fileset: Syntax error` the above throws on files with any of [-"'*] and it should be atomic.

**Comment by josephlou5** [comment]
> Oh, and we'll need to decide if `--raw-path=<PATH>` or `raw:<PATH>` means `cwd:quote(<PATH>)`, `root:quote(<PATH>)`, or `quote(<PATH>)`. For scripting, `root:quote(<PATH>)` is probably the most convenient, but `quote(<PATH>)` works better in interactive use.

IIUC, `quote(<PATH>)` will trigger the default fileset pattern kind, which is `cwd-prefix-glob:`. So that doesn't make sense here right?

FWIW, I don't think `root:quote(<PATH>)` is the "best" for scripting. Paths are most convenient to provide when relative to the working directory since your shell can give you tab completion or other features such as highlighting valid filenames. It sounds more tedious to let my script accept a path, but then have to resolve it relative to `jj root` before passing it to `raw:`. Therefore I think `cwd:quote(<PATH>)` is the most reasonable default, and `raw` can perhaps accept other modifiers such as `prefix-raw:` (or `raw-prefix:`?), `raw-i:`, and `root-raw:`.

> Other use casses:
> ```shell
> find -name \*.cc -o -name \*.hh -maxdepth 2 | xargs jj file track --raw --
> ```
> The above becomes hard when any kind of per file handling is required, it becomes close to impossible when template syntax specific escaping rules must be followed.

I don't think jumping straight to "(close to) impossible" is correct. In this specific example you can pass `-n 1` to make `xargs` run once per argument. This is of course not ideal (not the same operation, for example), but still very possible. Google also suggests a "printf hack" with `xargs sh -c 'eval "jj file track $(printf "--raw-path=%q" "$@")"' --` or by using `parallel -X`.

> No, just stick to what _every_ other unix tool does. The implementation should be this:
> ```shell
> # jj file track --raw -- *
> jj_file_track_raw() ( for f in "$@" ; do f=$(readlink -f -- "$f") && jj file track -- "file:$f" ; done )
> ```
> Except for the `Error: Failed to parse fileset: Syntax error` the above throws on files with any of [-"'*] and it should be atomic.

This is not correct; pretty sure `jj file track` accepts directory paths too, so restricting it to `file:` is wrong.

**Comment by yuja** [comment]
I have no idea why prefixing each path argument is so tedious. I agree that escaping strings in fileset/revset syntax isn't trivial, though.

This is also related: https://github.com/jj-vcs/jj/issues/3809

**Comment by dzaima** [comment]
> For interactive use, `jj file track '*'` can be used instead of letting the shell expand `*`.

There's of course the obvious problem of that still being an entirely-different feature from shell globbing, meaning missing out on special features shells may have for globbing, needing two levels of quoting for like `'"*.{cpp,hpp}"'`, not being able to tab-expand the glob to then manually remove certain files, etc etc etc, from forcing using a reinvented wheel when the regular thing could very trivially work perfectly fine.

It just feels extremely odd to me to not have some way of selecting files such that you can copy _exactly_ what you have in a bog-standard `rm <arbitrary shell file argument generation>` and pass that file list to jj instead (e.g. if you decide to, instead of removing the files, squash them to another revision instead), without forcing unnecessary jj-isms or praying that the files won't be wrongly-interpreted if parsed as filesets.

**Comment by tim-janik** [comment]
Ok, suppose we want to track or untrack files / directories which contain special chars like [-:  '*\\"~[\]] etc. The file names are provided via stdin so we use `find ... | xargs jj file track -- ` or via script / shell function arguments, so we use `jj file track -- "$@"`. Both approaches currently break as soon as special chars are encountered with `Error: Failed to parse fileset: Syntax error`, we need a single argument (or env var) that enables passing a list of absolute or cwd relative paths literally to jj commands.

FWIW, there is precedence for people needing support for this in other VCSs as well:

```sh
man git | fgrep -A1 literal-pathspecs
       --literal-pathspecs
           Treat pathspecs literally (i.e. no globbing, no pathspec magic). This is equivalent to setting the GIT_LITERAL_PATHSPECS environment variable to 1.
```

> I have no idea why prefixing each path argument is so tedious.

Per path prefixing breaks `jj file track --literal-pathspecs -- "$@"` and it breaks `find ... | xargs jj file track --literal-pathspecs --`
I am happy for you to show me how to make both examples work in a simple (non tedious) way.

> Google also suggests a "printf hack" with xargs sh -c 'eval "jj file track $(printf "--raw-path=%q" "$@")"' -- or by using parallel -X.

Using `printf %q` will emit shell-valid escapes that jj's string grammar rejects, its output isn't jj-parseable and %q is bash-only anyway, for example:

```sh
touch -- '--weird*file"'\'\! && jj file track $(printf file:%q --weird\*file\"\'\!)
Error: Failed to parse fileset: Syntax error
```

> This is not correct; pretty sure jj file track accepts directory paths too, so restricting it to file: is wrong.

Thank you for correcting me. I really made an effort to come up with a robust use of the current fileset syntax for a shell script and still failed. And that is exactly my point, I fail to see a simple way to currently use pathnames with jj commands without running into syntax errors or semantic errors with the existing fileset functionality.

This might just be a skill issue on my side, so I am happy for *ANYONE* to actually show me, how to make `find ... | xargs jj file track -- ` and `jj file track -- "$@"` work with arbitrary file names (valid unicdoe only).

There is also an error-reporting side to this that matters for scripts. Today, if a script passes a path that does not exist (typos), jj says nothing and exits with 0. I.e. `jj file track missing-file` silently succeeds even though no file was tracked, so a script has no way to notice the user of any mistake.
Special characters only make it worse, the script either gets that same silent success, or a misleading fileset syntax error that blames the quoting rather than the missing file. Only when jj knows a path is literal, could it actually report useful success or a helpful error message, e.g. `Error: path does not exist: missing-file`.
So a flag like --literal-pathspecs (globally supported) would not just make jj commands robust for arbitrary pathnames, it would also enable accurate feedback for wrong file paths.

**Comment by dzaima** [comment]
As a side-note of an alternative option if it gets decided that a flag for controlling this behavior will not be added - it would effectively work for scripting purposes to have a `jj util fileset-from-literal-file-paths-or-whatever <...>`, that generates a valid fileset expression that matches the passed-in args doing whatever escaping/quoting necessary and merges multiple paths with ` | `s, with which you could do `jj file track "$(jj util fileset-from-literal-file-paths -- **/*.{cpp,hpp})"`.

Would have the advantage that then you can do some fileset logic around the input paths if desired as it's just a substring now, but the obvious downside of it being massively-more verbose to the point that it'd still be hard to convince anyone to use it outside of really-long-term-use scripts, still leaving many places and shell usage praying that files parse to themselves as filesets.

**Comment by yuja** [comment]
> jj util fileset-from-literal-file-paths-or-whatever <...>

Yes, something like `jj util format-fileset`/`revset` seems useful.

```sh
$ jj util format-fileset -t 'root:%s' 'foo' 'bar' ...
(root:"foo") | (root:"bar")
$ jj util format-revset -t 'bookmarks(%s) & %r' 'foo' 'x&y' 'bar' 'z'
(bookmarks("foo") & (x&y)) | (bookmarks("bar") & (z))
```

For one-off stuff, `raw:`/`--raw-path` would be convenient.

**Comment by josephlou5** [comment]
> I am happy for _ANYONE_ to actually show me, how to make `find ... | xargs jj file track -- ` and `jj file track -- "$@"` work with arbitrary file names (valid unicdoe only).

I wrote https://gist.github.com/josephlou5/9f97d9d072a0434300e55b6d25b65bbe for this since it turned out to be fairly complicated. There are bash and fish functions, though the bash version is "better" since it handles control characters in filenames (but hopefully that's rare...). I also have versions that work with stdin piping, which should work well with `xargs`.
